File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer>list=new ArrayList<>();
            Stack<TreeNode> st=new Stack<>();
            
            TreeNode curr=root;
            
            while(curr!=null || !st.empty())
            {
                    if(curr!=null)
                    {
                            st.push(curr);
                            curr=curr.left;
                    }
                    else
                    {
                            TreeNode temp=st.peek().right;
                            if(temp == null)
                            {
                                    temp=st.peek();
                                    st.pop();
                                    list.add(temp.val);
                                    while(!st.empty() && temp==st.peek().right){
                                            temp=st.peek();
                                            st.pop();
                                            list.add(temp.val);
                                    }
                            }else
                            {
                                    curr=temp;
                            }
                    }
            }
            return list;
    }
}

File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int>preorder;
            stack<TreeNode*>st;
            //l r root
            if(root==NULL) return preorder;            
            
            st.push(root);
            while(!st.empty()) 
            {
                root =st.top();
                    st.pop();
                    
                    
                    if(root->left!=NULL){
                            st.push(root->left);
                    }
                    if(root->right!=NULL){
                            st.push(root->right);
                    }
                    preorder.push_back(root->val);
            }
            reverse(preorder.begin(),preorder.end());
            return preorder;
    }
};